graph: -f total-avg,self-avg,total-max,total-min,self-max,self-min in tui,cmds

junyeong0619 opened the pull request:
I'm sorry to make new PR. I pushed force to my master branch. So the branch's history completely changed and cannot reopen old PR. Thanks for your advice and it really helped me.

holyshitcode commented:
> Please split the changes in graph and tui. I think it's better to add the feature to the graph command first. And the second commit adds tui support. Also please update the documentation for each command in the same commit. Finally you need to add a test case for graph command.

I tried to split commit specifically. And add description on Document on each cmds,tui.

junyeong0619 commented:
> To run test, you can run the script with options and arguments. Please check `./runtest.py -h` for help.
> 
> I usually run it with the following options. The argument can be test number or name.
> 
> ```
> $ cd tests
> $ ./runtest.py -pvO2 187
> ```
> 
> Or you can run the same using `make` at the top level directory.
> 
> ```
> $ make runtest TESTARG="-pdO2 graph_field"
> ```

I tried to create a comparison test. I modified the fields that use nr_calls to check against zero in tui and cmds. Also, I aligned the elements in graph_field according to the layout defined in field.h.

junyeong0619 commented:
I believe I have addressed all the points from your review. Please kindly review it!

namhyung commented on cmds/graph.c:
```diff
@@ -70,6 +70,55 @@ static void print_addr(struct field_data *fd)
+static void print_graph_total_avg(struct field_data *fd)
+{
+	struct uftrace_graph_node *node = fd->arg;
+	uint64_t d;
```
Please add an empty line after declaration.

namhyung commented on cmds/graph.c:
```diff
@@ -70,6 +70,55 @@ static void print_addr(struct field_data *fd)
+	if (node->nr_calls > 0) {
+		d = node->time / node->nr_calls;
+		print_time_unit(d);
+	}
```
You should print something even if nr_calls == 0.  Something like this.

```c
uint64_t d = 0;

if (node->nr_calls)
    d = node->time / node->nr_calls;

print_time_unit(d);
```

namhyung commented on cmds/graph.c:
```diff
@@ -70,6 +70,55 @@ static void print_addr(struct field_data *fd)
+	if (node->nr_calls > 0) {
+		d = (node->time - node->child_time) / node->nr_calls;
+		print_time_unit(d);
+	}
```
Ditto.

namhyung commented on doc/uftrace-graph.md:
Please squash this change into the previous commit.

namhyung commented on cmds/tui.c:
```diff
@@ -291,6 +289,56 @@ static void print_graph_addr(struct field_data *fd)
+	if (node->nr_calls > 0) {
+		d = node->time / node->nr_calls;
+		print_time(d);
+	}
```
Please see my comment in cmds/graph.c

namhyung commented on cmds/tui.c:
```diff
@@ -291,6 +289,56 @@ static void print_graph_addr(struct field_data *fd)
+	if (node->nr_calls > 0) {
+		d = (node->time - node->child_time) / node->nr_calls;
+		print_time(d);
+	}
```
Ditto.

namhyung commented on doc/uftrace-tui.md:
Please squash this change into the previous commit.

namhyung reviewed (commented):
Overall, looks good to me, thanks!  A few nitpicks in the commit messages.

* Format the commit message properly.  Some line is too long, and some are unnecessarily indented.
* Move documentation changes into actual changes.
* Add an example output when you add new options or fields

junyeong0619 commented:
> Overall, looks good to me, thanks! A few nitpicks in the commit messages.
> 
> * Format the commit message properly.  Some line is too long, and some are unnecessarily indented.
> * Move documentation changes into actual changes.
> * Add an example output when you add new options or fields

Is this the way you intended the commits to be squashed? I’ve taken care of the other requested changes as well.
Thank you for your feedback. Please let me know if any further adjustments are needed.

namhyung reviewed (commented):
Thanks, I like the commits.  But the last point is that you need to indent code or shell output in the commit message by 2 or more spaces so that it can be distinguished.

junyeong0619 commented:
> Thanks, I like the commits. But the last point is that you need to indent code or shell output in the commit message by 2 or more spaces so that it can be distinguished.

I apologize for the oversight. While I had addressed this issue in the cmds part, I unintentionally missed applying it in the tui part.
I have now corrected the commit message and header alignment accordingly.

namhyung reviewed (approved):
LGTM, thanks a lot!
